src/de/deserializer.rs

```diff
@@ -0,0 +1,432 @@
+/// Options are assumed to be `Some` if there is any remaining data to be deserialized. If there is
+/// no remaining data, an option is assumed to be `None`
+///
+/// ### Sequences
+/// Sequences are assumed to be prefixed by the number of bytes that the entire sequence
+/// represents, stored as a variable integer. The length of the sequence is not known until the
+/// sequence has been deserialized because elements may have variable sizes.
+///
+/// ### Tuples
+/// Tuples are used as a special case of `sequence` where the number of elements, as opposed to the
+/// size of the binary data, is known at deserialization time. These are used to deserialize
+/// at-most N elements.
+///
+/// ### Other Types
+/// Structs, enums, and other variants will be mapped to either a `tuple` or a `sequence` as
+/// appropriate.
+///
+/// Other types are explicitly not implemented and there is no plan to implement them.
+use serde::de::{DeserializeSeed, IntoDeserializer, Visitor};
+use varint_rs::VarintReader;
+
+#[derive(Debug)]
+pub enum Error {
+    /// The feature is not designed to ever be supported.
+    WontImplement,
+
+    /// A custom deserialization error occurred.
+    Custom,
+
+    /// An invalid string was encountered, where UTF-8 decoding failed.
+    BadString,
+
+    /// An invalid boolean was encountered, which did not use "0" or "1" to encode its value.
+    BadBool,
+
+    /// There was not sufficient data to deserialize the required datatype.
+    InsufficientData,
+}
+
+impl serde::de::Error for Error {
+    fn custom<T: core::fmt::Display>(_msg: T) -> Self {
+        crate::error!("{}", _msg);
+        Error::Custom
+    }
+}
+
+impl core::fmt::Display for Error {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Error::WontImplement => "This feature won't ever be implemented",
+                Error::Custom => "Custom deserialization error",
+                Error::BadString => "Improper UTF-8 string encountered",
+                Error::BadBool => "Bad boolean encountered",
+                Error::InsufficientData => "Not enough data in the packet",
+            }
+        )
+    }
+}
+
+/// Deserializes a byte buffer into an MQTT control packet.
+pub struct MqttDeserializer<'a> {
+    buf: &'a [u8],
+    index: usize,
+}
+
+impl<'a> MqttDeserializer<'a> {
+    /// Construct a deserializer from a provided data buffer.
+    pub fn new(buf: &'a [u8]) -> Self {
+        Self { buf, index: 0 }
+    }
+
+    /// Attempt to take N bytes from the buffer.
+    pub fn try_take_n(&mut self, n: usize) -> Result<&'a [u8], Error> {
+        if self.len() < n {
+            return Err(Error::InsufficientData);
+        }
+
+        let data = &self.buf[self.index..self.index + n];
+        self.index += n;
+        Ok(data)
+    }
+
+    /// Pop a single byte from the data buffer.
+    pub fn pop(&mut self) -> Result<u8, Error> {
```

Comment:
Maybe if this is

```suggestion
    pub fn pop(&mut self, length: usize) -> Result<&'a [u8], Error> {
```

the repeated `pop()?; pop()?;...` below become a bit shorter.  And if the compiler isn't really smart it won't be able to merge all the early-return paths and length check/increment into one. This might produce less code.

Or even 

```suggestion
    pub fn pop<const N: usize>(&mut self) -> Result<&'a [u8], Error> {
```

Reply:
The `uXX::from_bytes()` functions take an array, not a slice. Because of that, we can't ask for an array back here.

Reply:
Why not?

Doesn't this work?

```rust
pub fn pop<const N: usize>(&mut self) -> Result<[u8; N], Error> {
...
    let mut r = [0u8; N];
    r.copy_from_slice(&buf[..N]);
    r
}
```

Reply:
We can't do this because `u32::from_be_bytes()` takes a `[u8; 4]` array and not a `&[u8]` slice

Reply:
Look at the `pop` signature. Works just fine.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9760c7047ec91616d3e4d41d44c1e5e0

Reply:
Er even better:

Reply:
I didn't know `&[u8]` had implementations for `TryInto<[u8; N]>` - I tried searching the TryFroms but must have been looking in the wrong place. Cleaned it up - thanks for the suggestion.